File: src/vct/col/ast/Hole.java
package vct.col.ast;

public class Hole extends ASTNode {

  private ThreadLocal<ASTNode> match=new ThreadLocal<ASTNode>();
  
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
   }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 
 
  public ASTNode get(){
    return match.get();
  }

  public boolean match(ASTNode n){
    match.set(n);
    return true;
  }
}

File: vercors/src/main/java/vct/col/ast/stmt/composite/Hole.scala
package vct.col.ast.stmt.composite

import vct.col.ast.generic.ASTNode
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

class Hole(private[this] val nodes:ThreadLocal[ASTNode]) extends ASTNode with VisitorHelper {
  def get : ASTNode = nodes.get
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def `match`(ast:ASTNode) = { nodes.set(ast); true }

  override def debugTreeChildrenFields(): Iterable[String] = Seq("nodes")
  override def debugTreePropertyFields(): Iterable[String] = Seq()
}